go/connector/run.go

```diff
@@ -167,29 +142,55 @@ func runCommand(
 	var cmd = exec.Command(args[0], args[1:]...)
 	var fe = new(firstError)
 
-	// Copy |writeLoop| into connector stdin.
-	wc, err := cmd.StdinPipe()
-	if err != nil {
-		return fmt.Errorf("StdinPipe: %w", err)
-	}
-	go func() {
-		defer wc.Close()
-		fe.onError(writeLoop(wc))
-	}()
-
-	var stderrForwarder = ops.NewLogForwardWriter("connector stderr", logrus.InfoLevel, logger)
-
-	// Decode and forward connector stdout to |output|, but intercept a
+	// Decode and forward connector output from socket to |output|, but intercept a
 	// returned error to cancel our context and report through |fe|.
 	// If we didn't cancel, then the connector would run indefinitely.
-	cmd.Stdout = &writeErrInterceptor{
+	var outputInterceptor = &writeErrInterceptor{
 		delegate: output,
 		onError: func(err error) error {
 			fe.onError(err)
 			cancel() // Signal to exit.
 			return err
 		},
 	}
+
+	// Routine waiting for a connection from connector to be established to start
+	// streaming input/output
+	var conn net.Conn
+	go func() {
+		// Accept connection on the socket
+		var err error
+		conn, err = socket.Accept()
+		if err != nil {
+			fe.onError(fmt.Errorf("accepting connection from socket: %w", err))
+			return
+		}
+
+		// Copy |writeLoop| into socket
+		go func() {
+			fe.onError(writeLoop(conn))
+		}()
+
+		// Read from socket connection and delegate to output through the error interceptor
+		go func() {
+			var _, err = io.Copy(outputInterceptor, conn)
+			fe.onError(err)
+		}()
+	}()
+
+	// Give ready signal to connector, specifying the socket type and address
+	cmdStdin, err := cmd.StdinPipe()
+	if err != nil {
+		return fmt.Errorf("StdinPipe: %w", err)
+	}
+	_, port, err := net.SplitHostPort(socket.Addr().String())
+	if err != nil {
+		return fmt.Errorf("splitting socket address: %w", err)
+	}
+	cmdStdin.Write([]byte(fmt.Sprintf("tcp %s:%s\n", DockerHostInternal, port)))
```

Comment:
Would it be possible to provide this as an argument to the connector rather than sending it over `stdin`? I'm not sure if it's needed to be used as a "ready" signal here with the section above waiting to receive a connection on the socket before doing anything with it. Using an argument instead of `stdin` seems simpler to me, but not an absolute must-do (especially if it doesn't work).